### ISSUE TITLE
(how) are standard notifications being consolidated?

### ISSUE THREAD

**Issue by JChipault** [USGS-WiM/whispersservices_django#361]
Kim created event 170863 on 3/23 and then I edited 170863 on 3/23. I was expecting 2 notifications: 1 saying that Kim created it and 1 saying that Jenny updated it.  But I only received 1 notification total (indicating that Kim created it). 

If we were basic partner users and not part of the development group, we think we would expect 1 email with all the notifications pertaining to an event consolidated (so in this case, 1 email indicating that event 170863 was both created and updated on 3/23). Mainly because that would be tidy, but perhaps that's a big pain to code?

In lieu of that, we would expect a notification for each unique trigger. So, in this example, an email that a new event had been entered by Kim and an email that it had been changed by Jenny.

But, that said, if we can't indicate what type of field was updated then getting a new notification for every separate update might get crazy. Would 5 changes to the database all made by one person in one session in one event be seen as 5 update triggers or 1? It will seems like a glitch to the user if they receive 5 identical emails that an event was updated.  

Overall, if you are consolidating notifications, can you explain how that's being done? Kim and I can't find any specifics in our notes of how consolidation was going to go.

**Comment by aaronstephenson** [comment]
Notifications are not consolidated. The reason you are seeing a notification for "created" but not "updated" is because both actions occurred on the same day. When I search for updated events, I specifically exclude events created that same day. Here's the code:
```
    yesterday = datetime.strftime(datetime.now() - timedelta(1), '%Y-%m-%d')
    new_events = Event.objects.filter(created_date=yesterday)
    updated_events = Event.objects.filter(modified_date=yesterday).exclude(created_date=yesterday)
```
The main reason for this is because `modified_date` is populated when the object is created (exactly like the behavior of the created date and modified date attributes for files in Windows). Additionally, nothing in the notifications reference spreadsheet mentions a situation where an event would be both created and modified in a single day. Every example says either "created" (or "new") or "updated", not "created and updated":
![image](https://user-images.githubusercontent.com/8713712/77583281-d4b1ac00-6eae-11ea-8cdb-b32c55536699.png)

As for your question about multiple updates to an event, only a single notification would result. I don't have my notes in front of me, but I remember a consensus decision in a meeting several months back that, to reduce the number of "duplicate" notifications each day (such as a created or updated event appearing in "ALL Events" and "Your Events" and "Organization Events"), only the "highest" notification would be created and the others ignored (e.g., create one notification for "ALL Events" but not "Your Events" or "Organization Events"). The way I'm checking for "duplicates" is comparing the combination of recipient ID and event ID values of each notification.

Just an aside, but the scenario you describe (5 different updates to an event in one session) is another reason why querying and consolidating updates to be included in the notification text will be a lot of work.

**Comment by JChipault** [comment]
FYI - we talked about this as a group today and have some thoughts but are doing more testing and making sure we're all on the same page at NWHC before we chat with you about it. Thanks for providing the information above!

**Comment by JChipault** [comment]
@aaronstephenson Regarding this part, "to reduce the number of "duplicate" notifications each day (such as a created or updated event appearing in "ALL Events" and "Your Events" and "Organization Events"), only the "highest" notification would be created and the others ignored (e.g., create one notification for "ALL Events" but not "Your Events" or "Organization Events"). "

can we make the hierarchy match what users see on the dashboard, with "Your Events" as the "highest" ranking instead of "All Events"? The "All Events" email text just provides the organization that made the change, whereas the other standard notifications list an individual's name (because you're close to those events and know the org) so the "All Events" alert is the least informative of the standard notifications and we'd prefer other alerts to go through if they apply.

![image](https://user-images.githubusercontent.com/49287076/78268604-08ec2480-74ce-11ea-91d5-402c0964ad3b.png)

**Comment by aaronstephenson** [comment]
Yes, I can make that change. It is in the current order because that's the order they are listed in the Notifications reference spreadsheet.

**Comment by aaronstephenson** [comment]
Changed in b52a4dd5

**Comment by JChipault** [comment]
After more testing, NWHC talked again and we are wondering if it would be possible to receive separate updated event notifications for each user that made an edit (i.e. 'source') for "Your Events" "Your Organization's Events" and "Your Collaborator Events".  

So if event 200501 were edited today by Ali (3 edits), Neil (1), Jenny (5), and Kim (2) I would get 4 emails tomorrow:
User Ali Rahama has updated your organization's WHISPers Event 200501 on 4/3/2020.
User Neil Baertlein has updated your organization's WHISPers Event 200501 on 4/3/2020.
User Jenny Chipault has updated your organization's WHISPers Event 200501 on 4/3/2020.
User Kim Miller has updated your organization's WHISPers Event 200501 on 4/3/2020.

Since the All Events standard notification and the custom notifications just list organization in the message to users, we'd like those notifications consolidated by organization if possible. So in the above example, since all those people are from NWHC, a user with an All Events alert would simply receive:
WHISPers Event 200501 in Modoc, CA, USA was updated on 2020-04-03 by U.S. Geological Survey - National Wildlife Health Center.

Is this do-able? The goal is to deliver as much information to the user as possible (e.g., if it's your owned event, you might want to know who exactly was in there editing) without overwhelming users with duplicative emails.

**Comment by BlakeDraper** [comment]
@JChipault @nbaertlein Aaron was able to do some lifting and complete this consolidation change for you guys. We are still testing it on our dev environments but expect it to be on test shortly. 

We also have a question for you guys: for the organization-based consolidated notifications for Custom and All Events, should the `Source` field value be the organization? In all other notifications the source is a specific user whose actions initiated/triggered the notification. Since these ones in question are consolidated by org, there is no way to report the source aside from it being the organization (well, there are other ways, but this is the way we prefer). Do you approve of this approach?

**Comment by nbaertlein** [comment]
@BlakeDraper Organization as source for custom and all events sounds good to us.

**Comment by BlakeDraper** [comment]
@nbaertlein @JChipault Aaron has published these requested changes to the test environment for you all to work with. Hopefully this brings us to a workable notifications release to production.

**Comment by JChipault** [comment]
Also received "All Events" notification for 170872 that had Ali's and Kim's changes noted with NWHC as org… We were expecting the Your Events notification to have trumped this All Events notification, but maybe now that Your/Your Org/Your Collaborator notifications have a source of user and All/Custom have a source of organization, these are in two different pools so we should expect to get both Your and All?? But when I created event 170879 on 4/20, on 4/21 I only got a Your Events notification. Just checking in on this.

**Comment by JChipault** [comment]
Any possibility of starting new lines of text for each different update instead of having them all smashed together in one paragraph? We know this is not a bug so we might need to accept as is.

**Comment by JChipault** [comment]
And last bug....
On 4/22 I added one sick arroyo toad to 170861, which should have triggered "Your Events" (and maybe "All Events"??... see above) and my "California" custom trigger, but I received no notifications about this overnight. The fact that this species update didn't trigger any notifications, paired with the fact that I never got notified about the species update I made on 4/20, makes me wonder if there's something not triggering with regard to the species being updated.

**Comment by JChipault** [comment]
Today at 10:30a, I added a species diagnosis to a Achilles Tang in event 170861 (an event in California) and received no custom notification at 11a. This is probably related to my last comment, but throwing it here so you have more data. 

Now I'll stop testing until you tell me to restart testing.

**Comment by aaronstephenson** [comment]
Thanks for all the details @JChipault it really helps. Regarding the last comment about Achilles Tang, that wouldn't show up until tomorrow; the checks are still looking at changes made yesterday.

For testing purposes, would you like the change detection to be for today instead?

**Comment by aaronstephenson** [comment]
OK, everything reported here should be fixed by e931ab2c.

@JChipault @nbaertlein  please resume testing at your convenience.

Note that the change detection is still looking for changes made the previous day, please let me know if you want that changes to the current day for the duration of testing.

**Comment by JChipault** [comment]
Good news! I added a species dx yesterday and received a Your Events and Custom notification today as expected and all looks good. 

I would prefer to be able to test adds/edits/deletes throughout the day but I don't know how much recoding that is for you.

**Comment by aaronstephenson** [comment]
Good to hear!

And the date check is a super simple change, I'll do it right now.

**Comment by aaronstephenson** [comment]
Done. Changes will be detected for the current day.

During testing, Standard Notifications will be processed on the hour at 3,10,12,14,16 and Custom Notifications will be processed at 4,11,13,15,17 (all times Central).

**Comment by JChipault** [comment]
Any chance we can get the cause and basis for species diagnoses interpreted? They just say "1"

![image](https://user-images.githubusercontent.com/49287076/80397175-5e241780-887b-11ea-948d-821ee3d572c9.png)

**Comment by aaronstephenson** [comment]
Absolutely. Sorry, oversight when I implemented all the other lookups.

**Comment by aaronstephenson** [comment]
Implemented in 445b18f0

**Comment by JChipault** [comment]
Thanks, that looks good now

**Comment by JChipault** [comment]
Next suite... at 1045 today I logged in as the abilenetest user (first name: Abilene, last name: Zoo) and added A) the species diagnosis of Duck Viral Hepatitis suspect and B) location comment of "adding a comment to test". At 12p, I got two Standard notifications about this change:
![image](https://user-images.githubusercontent.com/49287076/80413343-f7abf300-8894-11ea-947e-aa75df637404.png)

Few things:
1) I received both a "your events" and a "all events" notifications... I was only expecting a "your events" but maybe there's some logic here I'm not following
2) On the "your events", it says that "User Jenny Chipault has updated..." and I was expecting "User Abilene Zoo has updated..."
3) There's no mention of the added comment. I went back in to this event at 11:40a and added an event comment to this event as jchipault and never got a notification about that comment either. So adding comments doesn't seem to be triggering notifications. Furthermore, the custom notifications that I've received about this event don't mention these two added comments either.
4) Maybe this is too hard to do so feel free to pass on this one but I'm wondering if it'd be easy to relay that this was enter as Duck Viral Hepatitis suspect (this doesn't relay the "suspect" part)

**Comment by JChipault** [comment]
Ran another round of tests with results coming in at 3p. Overall, pretty darn good! I'm impressed. I made a lot of changes to 170864 and the only change that didn't come through was the fact that I added an event organization... but maybe that's to be expected based on what tables are getting pinged?

**Comment by BlakeDraper** [comment]
@JChipault we are glad to hear that! Yes,event org is not among the models checked. 

Event `modified_by` and `modified_date` fields are updated when the following models are saved (creates and updates) or deleted:

- Event
- EventLocation
- EventLocationContact
- EventLocationFlyway
- LocationSpecies
- EventDiagnosis
- SpeciesDiagnosis
- SpeciesDiagnosisOrganization
- Comment (only when attached to Event, EventLocation, or EventEventGroup)

**Comment by aaronstephenson** [comment]
@JChipault I can include Event Organization, it should only take an hour or so.

**Comment by aaronstephenson** [comment]
Saves and deletes of Event Organization (and also Event Contact) now update Event `modified_date` and `modified_by` fields, and are also included in the updates list in notifications. Implemented in 7a15b8e7

**Comment by JChipault** [comment]
Doing more testing today. It's exciting to be so close to done; and so awesome to have the fields that were changed noted within the notifications! 

Just wanted to make sure the quirks reported in the "Next suite..." post above is still on the radar to be assessed.

**Comment by aaronstephenson** [comment]
Yes still on the radar.

**Comment by JChipault** [comment]
Looks like you over-achieved on the addition of the event org... we got an extra "was"
![image](https://user-images.githubusercontent.com/49287076/80524875-f68eca80-8955-11ea-954c-40e0884118fd.png)

**Comment by JChipault** [comment]
I'm getting notifications about event comments via "All Events" and "Your Events" now... will do some more testing today to check if location comments come through and if custom notifications serve up comments now.

**Comment by JChipault** [comment]
I added both an event comment and a location comment to 170880 and got the notification below. Looks like location comments are not coming through.
![image](https://user-images.githubusercontent.com/49287076/80738289-61690e80-8ada-11ea-8e82-541cfd5027ee.png)

**Comment by JChipault** [comment]
Same results regarding comments on the custom notifications
![image](https://user-images.githubusercontent.com/49287076/80746438-82842c00-8ae7-11ea-876a-cfe7b71ce287.png)

**Comment by aaronstephenson** [comment]
All bugs reported here should now be fixed with latest commit 9ba5dfc2
